README.md
<h1 align="center">
  <img width="350px" alt="Idena Coacher Node Monitor Tool" Title="IDENA Coacher - Node Monitor Tool" src="https://raw.githubusercontent.com/ltraveler/ltraveler/main/images/idena_coacher.png"><br/>
  Idena Coacher — Node Monitor Tool
</h1>
<p align="center">🖥️ Manage locally your IDENA node without having to run Idena desktop client with enabled port forwarding.<br> <b><i>★ Mining status change</i></b> ☆ <b><i>Transfer coins</i></b> ☆ <b><i>Stake replenishment</i></b> ☆ <b><i>Sync status</i></b> ☆ <b><i>Available balance ★</i></b></p>

<p align="center"><a href="https://github.com/ltraveler/idena-coacher/releases/latest" target="_blank"><img src="https://img.shields.io/github/v/release/ltraveler/idena-coacher?style=for-the-badge&logo=none" alt="idena runner latest version" /></a>&nbsp;<a href="https://wiki.ubuntu.com/FocalFossa/ReleaseNotes" target="_blank"><img src="https://img.shields.io/badge/Kali-20.04(LTS)+-00ADD8?style=for-the-badge&logo=none" alt="Kali minimum version" /></a>&nbsp;<a href="https://github.com/ltraveler/idena-runner/blob/main/CHANGELOG.md" target="_blank"><img src="https://img.shields.io/badge/Build-Stable-success?style=for-the-badge&logo=none" alt="idena-go latest release" /></a>&nbsp;<a href="https://www.gnu.org/licenses/quick-guide-gplv3.html" target="_blank"><img src="https://img.shields.io/badge/license-GPL3.0-red?style=for-the-badge&logo=none" alt="license" /></a>&nbsp;<a href="https://github.com/ltraveler/idena-coacher/blob/main/README.ru-RU.md" target="_blank"><img src="https://img.shields.io/badge/readme-РУССКИЙ-orange?style=for-the-badge&logo=none" alt="Скрипт Idena ARMer" /></a></p>

<p align="center"><br>
  <img alt="Idena Coacher Node Monitor Tool" Title="IDENA Coacher - User Interface" src="https://raw.githubusercontent.com/ltraveler/ltraveler/main/images/IDENA_Coacher_Monitor_Tool_UI.jpg">
</p>

## 🚀&nbsp; Quick Start
*1. Clone Idena Coacher to the **home directory of the user** that runs idena-go node client.*
```
cd /home/%username%
git clone https://github.com/ltraveler/idena-coacher.git
cd idena-coacher
```
*$username - the name of the user that runs idena-go client<br><br>
*2. Make idena_coacher.sh executable*
```
chmod +x idena_coacher.sh
```
*3. Originally IDENA Coacher has been developed as an additional TUI tool to manage nodes that have been deployed by [IDENA ARMer](https://github.com/ltraveler/idena-armer) or [IDENA Runner](https://github.com/ltraveler/idena-runner) scripts. Simply skip this step in case if you are willing to use it with the above-mentioned tools.<br><br>
In case if your node has been deployed customly you would have to change the default paths in the main script file.<br><br>
Please run the command `nano idena_coacher.sh` in the home folder of the script and edit the following lines:*
```
PRIVATE_PATH="/home/$username/idena-go/datadir/keystore/nodekey"
```
*Line #7:* **Path to the nodekey (private key) file**
```
API_PATH="/home/$username/idena-go/datadir/api.key"
```
*Line #9:* **Path to the node api.key file**
```
LOG_PATH="/home/$username/idena-go/datadir/logs/output.log"
```
*Line #10:* **Path to the idena-go log file**
```
export RPC_PORT="9009"
```
*Line #12:* **RPC Port of your node**

*4. Run the script.*
```
./idena_coacher.sh
```
## ⏳&nbsp; Update the script to the latest version

* **Enter to the script home directory and run `coacher_update.sh`**
```
cd /home/$username/idena-coacher/
./coacher_update.sh
```
**$username* - the name of the user that runs `idena-go` client


CHANGELOG.md
## 0.1.10 (Nov 03, 2022)

### Bugfix

- Small fix realted to IDENA ARMer '$PATH' change 

## 0.1.9 (Oct 27, 2022)

### Bugfix

- IDENA Coacher update script has been changed
- idena-go terminal output fix

## 0.1.8 (Oct 26, 2022)

### Changes

- IDENA Coacher update script has been added

## 0.1.7 (Oct 25, 2022)

### Bugfix

- Showing the whole stake instead of the replenishment part of it

## 0.1.6 (Oct 23, 2022)

### Changes

- Stake replenishment transaction

## 0.1.5 (Oct 23, 2022)

### Changes

- First released version


idena_coacher.sh
#!/usr/bin/bash
#
#Checking if req packages are installed
reqpkgs=('jq' 'dialog' 'nano' 'netcat-traditional' 'gettext')
for x in "${reqpkgs[@]}"; do
dpkg -s "$x" &> /dev/null
if [ $? != 0 ]; then
    echo -e "${LRED}Package $x is not instlled. Installing...${NC}"
    apt-get install -y $x; 
fi
done
#
TERM=linux
clear
#
ARMER_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
username=$(echo "$ARMER_DIR" | awk -F\/ '{print $3}')
cd /home/$username/idena-coacher
PRIVATE_PATH="/home/$username/idena-go/datadir/keystore/nodekey"
PRIVATE_KEY=$(cat "$PRIVATE_PATH")
API_PATH="/home/$username/idena-go/datadir/api.key"
LOG_PATH="/home/$username/idena-go/datadir/logs/output.log"
export API_KEY=$(cat $API_PATH)
export RPC_PORT="9009"
export RPC_HOST="http://localhost"

export node_address=$(envsubst < ./api/node_address | bash) 

node_status=$(envsubst < ./api/node_status | bash)
current_block=$(envsubst < ./api/current_block | bash)
highest_block=$(envsubst < ./api/highest_block | bash)
current_balance=$(envsubst < ./api/get_balance | bash)
current_stake=$(envsubst < ./api/get_stake | bash)

dna_status=$(envsubst < ./api/dna_status | bash)

function delayProgress {

                for ((i=0;i<=100;i+=10)); do echo $i; sleep 3; done | dialog --gauge "Please wait 30 seconds." 0 0

}

function importPkey {

IBOX_KEY=$(dialog --colors --backtitle "\Zb\Z5Your current key: \Zu\Z3$PRIVATE_KEY\Zn" --title "Private Key Import Dialog" --inputbox "Enter your nodekey?"  8 39  3>&1 1>&2 2>&3)
exitstatus=$?
if [ $exitstatus = 0 ] && [ -n "$IBOX_KEY" ] && [ "$PRIVATE_KEY" != "$IBOX_KEY" ]; then
#echo "Updating your Private Key to $PRIVATE_KEY"
    service idena stop
    echo "$IBOX_KEY" > "$PRIVATE_PATH"
    service idena start > /dev/null
    for ((i=0;i<=100;i+=10)); do echo $i; sleep 6; done | dialog --gauge "Please wait 60 seconds." 0 0
else
    dialog --title "Private key import" --clear --msgbox "Private key import aborted. No changes were made." 10 41
fi

echo "(Exit status: $exitstatus)"
	}

function showLog {

	dialog --colors --ascii-lines --tailbox "$LOG_PATH" 60 160

	}

function dashRefresh {

		PRIVATE_KEY=$(cat "$PRIVATE_PATH")
		export node_address=$(envsubst < ./api/node_address | bash)
		node_status=$(envsubst < ./api/node_status | bash)
		current_block=$(envsubst < ./api/current_block | bash)
		highest_block=$(envsubst < ./api/highest_block | bash)
		dna_status=$(envsubst < ./api/dna_status | bash)
		current_balance=$(envsubst < ./api/get_balance | bash)
		current_stake=$(envsubst < ./api/get_stake | bash)
	}

function sendDNA {
#
export transaction_type="0"
#
destination_address=""
transaction_amount=""
#
# open fd
exec 3>&1

# Store transaction details
export VALUES=$(dialog --colors --ok-label "Send" \
	  --backtitle "Measure payee address thrice and send once! \Z1\ZbATTENTION:\ZB\Zn \Z3\ZbThere is no way\Zn to cancel transaction after clicking \Z1Send\Zn button!\ZB" \
	  --title "$node_address" \
	  --form "\\Z1Current balance: \Z1\Zb$current_balance\ZB\Zn" \
15 70 0 \
	"Destination:" 1 1	"$destination_address" 	1 15 42 0 \
	"Amount to transfer:"    2 1	"$transaction_amount"  	2 23 34 0 \
2>&1 1>&3)
exitstatus=$?
# close fd
exec 3>&-
echo $VALUES
export destination_address=`echo $VALUES | awk '{print $1}'`
export transaction_amount=`echo $VALUES | awk '{print $2}'`
	
if [[ $exitstatus = 0 ]] && [[ -n "$destination_address" ]] && [[ "$destination_address" =~ ^0x[0-9a-fA-F]{40}$ ]] && [[ -n "$transaction_amount" ]] && [[ "$transaction_amount" =~ ^[0-9]+$ ]]; then

	transaction_hash=$(envsubst < ./api/send_dna | bash)
	for ((i=0;i<=100;i+=10)); do echo $i; sleep 3; done | dialog --gauge "Please wait 30 seconds." 0 0
	dialog --colors --title "Transaction has been sent" --clear --msgbox "\Zb\Z3Destination:\ZB\n\Zu\Z0$destination_address\Zn\n\n\Zb\Z3Amount:\ZB\n\Zu\Z0$transaction_amount\Zn\n\n\Zb\Z3Transaction hash:\ZB\n\Zu\Z0$transaction_hash\Zn" 12 80 
	dashRefresh

else

	dialog --colors --title "Oops, something went wrong..." --clear --msgbox "\Z1Please check one of the following conditions:\ZB\Zn\n\n1. You may have entered an incorrect payee \Z4iDNA\Zn address.\n2. You may have entered an incorrect \Z4iDNA\Zn amount.\n3. The \Z5ESC\Zn or \Z5Cancel\Zn button has been pressed." 10 80

fi

}
###
function replenishStake {
#
export transaction_type="22"
#
destination_address="$node_address"
transaction_amount=""
#
# open fd
exec 3>&1

# Store transaction details
export VALUES=$(dialog --colors --ok-label "Replenish" \
	  --backtitle "Measure payee address thrice and send once! \Z1\ZbATTENTION:\ZB\Zn \Z3\ZbThere is no way\Zn to cancel stake replenishment transaction after clicking \Z1Replenish\Zn button!\ZB" \
	  --title "$node_address" \
	  --form "\Z1Current balance: \Z1\Zb$current_balance\ZB\Zn" \
15 70 0 \
	"Destination:" 1 1	"$destination_address" 	1 15 42 0 \
	"Replenishment amount:"    2 1	"$transaction_amount"  	2 23 34 0 \
2>&1 1>&3)
exitstatus=$?
# close fd
exec 3>&-
echo $VALUES
export destination_address=`echo $VALUES | awk '{print $1}'`
export transaction_amount=`echo $VALUES | awk '{print $2}'`
	
if [[ $exitstatus = 0 ]] && [[ -n "$destination_address" ]] && [[ "$destination_address" =~ ^0x[0-9a-fA-F]{40}$ ]] && [[ -n "$transaction_amount" ]] && [[ "$transaction_amount" =~ ^[0-9]+$ ]]; then

	transaction_hash=$(envsubst < ./api/send_dna | bash)
	for ((i=0;i<=100;i+=10)); do echo $i; sleep 3; done | dialog --gauge "Please wait 30 seconds." 0 0
	dialog --colors --title "Transaction has been sent" --clear --msgbox "\Zb\Z3Destination:\ZB\n\Zu\Z0$destination_address\Zn\n\n\Zb\Z3Amount:\ZB\n\Zu\Z0$transaction_amount\Zn\n\n\Zb\Z3Transaction hash:\ZB\n\Zu\Z0$transaction_hash\Zn" 12 80 
	dashRefresh

else

	dialog --colors --title "Oops, something went wrong..." --clear --msgbox "\Z1Please check one of the following conditions:\ZB\Zn\n\n1. You may have entered an incorrect payee \Z4iDNA\Zn address.\n2. You may have entered an incorrect \Z4iDNA\Zn replenishment amount.\n3. The \Z5ESC\Zn or \Z5Cancel\Zn button has been pressed." 10 80

fi

}
#checking if idena-go client is running
if ! ( nc -zv localhost $RPC_PORT 2>&1 >/dev/null ); then
dialog --colors --title "Oops, something went wrong..." --clear --msgbox "\Z1The port \Z4$RPC_PORT\Zn \Z1is closed!\Zn\n\nPlease check if the \Zb\Z4RPC_PORT\Zn\ZB variable has the correct port number and your \Z4idena-go\Zn node client is successfully running." 10 80
clear
exit 1
fi
###
while [ 1 ]
do
CHOICE=$(
dialog --colors --clear --backtitle "\Zb\Z5Current block: \Zu\Z3$current_block\Zn \Zb\Z5Highest block: \Zu\Z3$highest_block\Zn \Zb\Z5Current Balance: \Zu\Z3$current_balance\Zn \Zb\Z5Current Stake: \Zu\Z3$current_stake\Zn \Zb\Z5Mining Status: \Zu\Z3$dna_status\Zn"  --title "Idena Node Management Tool by \ZB\Z5LTraveler\Zb\Zn" --menu " \Zb\Z3Node ID: \ZB\Z0$node_address\Zb\Zn  \Zb\Z3Sync Status: \ZB\Z0$node_status\Zb\Zn" 15 78 5 \
	"1)" "Show Private Key"  \
	"2)" "Import Private Key"  \
	"3)" "Terminal Layout"  \
	"4)" "Activate Mining" \
	"5)" "Deactivate Mining" \
	"6)" "Send IDNA" \
	"7)" "Stake Replenishment" \
       	"8)" "About IDENA Coacher" 3>&2 2>&1 1>&3	
)

if [ $? -gt 0 ]; then 
        break
fi

case $CHOICE in
	"1)")   
		dashRefresh
		dialog --colors --title "IDENA Node Wallet Info" --msgbox "\Zb\Z3Address:\ZB\n\Zu\Z0$node_address\Zn\n\n\Zb\Z3Private key:\ZB\n\Zu\Z0$PRIVATE_KEY\Zn" 10 80
	;;
	"2)")   
		importPkey
		dashRefresh
	;;

	"3)")   
		showLog
		dashRefresh
        ;;

	"4)")   
		#dialog --colors --msgbox "Under development" 20 78
		envsubst < ./api/mining_on | bash
		delayProgress
		dashRefresh
        ;;

	"5)")   
             	#dialog --colors --msgbox "Under development" 20 78 
		envsubst < ./api/mining_off | bash
		delayProgress
                dashRefresh
        ;;

	"6)")   
		#dialog --colors --msgbox "Under development" 20 78	
		sendDNA
		dashRefresh
        ;;
	"7)")   
		#dialog --colors --msgbox "Under development" 20 78	
		replenishStake
		dashRefresh
        ;;

	"8)")
		dialog --colors --title "IDENA Coacher by LTraveler: (\Zb\Z6t.me/ltrvlr\Zn\ZB)" --msgbox "\nBash implementation of the localhost node management tool with the following features:\n✧ \Z5Status\Zn ✧ \Z3\ZbMining ON/OFF\Zn\ZB ✧ \Z1Key change\Zn ✧ \Zb\Z4Send iDNA\Zn\ZB ✧\n\n✦ \Z7GitHub Repository:\Zn \Zb\Z7https://github.com/ltraveler/idena-coacher\Zn\n✧ \Z7Version:\Zn \Zb\Z70.1.5\Zn\ZB\n✦ \Z7Any donations are welcome:\Zn \Zb\Z70xf041640788910fc89a211cd5bcbf518f4f14d831\Zn\ZB\n✧ \Z7Telegram support:\Zn \Zb\Z7https://t.me/ltrvlr\Zn\ZB" 13 91
	;;

esac
done
clear
exit

